src/prefect/cli/flow_run.py

```diff
@@ -231,8 +247,21 @@ async def logs(
             # Update the number of logs retrieved
             num_logs_returned += num_logs_to_return_from_page
 
-            if len(page_logs) == LOGS_DEFAULT_PAGE_SIZE:
-                offset += LOGS_DEFAULT_PAGE_SIZE
+            if tail:
+                if offset != 0:
+                    offset = (
+                        0
+                        if offset < LOGS_DEFAULT_PAGE_SIZE
+                        else offset - LOGS_DEFAULT_PAGE_SIZE
+                    )
+                else:
+                    more_logs = False
             else:
-                # No more logs to show, exit
-                more_logs = False
+                if (
+                    num_logs_returned < user_specified_num_logs
+                    and len(page_logs) == LOGS_DEFAULT_PAGE_SIZE
+                ):
+                    offset += LOGS_DEFAULT_PAGE_SIZE
+                else:
+                    # No more logs to show, exit
+                    more_logs = False
```

Comment:
Some comments here explaining what is happening will be useful for future implementers :)